### PR TITLE
App Service: Web app & deployment slots

### PR DESCRIPTION
**Pull request by jianghaolu**

### CODE REVIEW

**Comment by jianghaolu** [comment]
The issues discussed are addressed, except the following:
- getPfxFromKeyVault
- getPfxPasswordFromKeyVault

Key vault access requires separate access policy to set on the vault, and creating a Key Vault data plane client, and query the data. It can be implemented in the near future.

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/AppServiceCertificate.java [hunk 1]
```diff
@@ -6,28 +6,183 @@
...
+
+    /**
+     * @return if the certificate valid
+     */
+    Boolean valid();
+
+    /**
+     * @return the raw bytes of .cer file
+     */
+    String cerBlob();
```
Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/AppServiceCertificate.java]: generally abbreviations in API names are discourages.,  --> certificateBlob()

Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/AppServiceCertificate.java]: generally abbreviations in API names are discouraged.,  --> certificateBlob()

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/DeploymentSlot.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+         */
+        interface Blank extends WithConfiguration {
+        }
+
+        /**
+         * A deployment slot definition allowing the configuration to clone from to be specified.
+         */
+        interface WithConfiguration {
+            /**
+             * Create the deployment slot with brand new site configurations.
```
Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/DeploymentSlot.java]: descriptions of methods are written in 3rd person declarative mood. i.e. "Creates ...." etc. Theer are multiople instances of this in these files.

Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/DeploymentSlot.java]: descriptions of methods are written in 3rd person declarative mood. i.e. "Creates ...." etc. There are multiple instances of this in these files.

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/DomainContact.java [hunk 3]
```diff
@@ -86,80 +86,200 @@
...
         interface WithFaxNumber<ParentT> {
             WithAttach<ParentT> withFaxNumber(String faxNumber);
         }
 
+        /**
+         * The stage of contact definition allowing organization to be set.
+         * @param <ParentT> the return type of the final {@link WithAttach#attach()}
+         */
         interface WithOrganization<ParentT> {
             WithAttach<ParentT> withOrganziation(String organziation);
```
Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/DomainContact.java]: typo --> withOrganization()

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/HostNameSslBinding.java [hunk 4]
```diff
@@ -35,78 +36,230 @@
...
+         * @param <ParentT> the return type of the final {@link WithAttach#attach()}
+         */
         interface WithSslType<ParentT> {
-            WithAttach<ParentT> withSniSSL();
-            WithAttach<ParentT> withIpBasedSSL();
+            /**
+             * Use Server Name Indication (SNI) based SSL.
+             * @return the next stage of the hostname SSL binding definition
+             */
+            WithAttach<ParentT> withSniSsl();
```
Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/HostNameSslBinding.java]: sincd the other one is called withIpBasedSsl() how about --> withSniBasedSsl() for this to be consistent and a bit clearer?

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/HostNameSslBinding.java [hunk 5]
```diff
@@ -35,78 +36,230 @@
...
+        }
+
+        /**
+         * The stage of hostname SSL binding definition allowing certificate information to be specified.
+         * @param <ParentT> the return type of the final {@link WithAttach#attach()}
+         */
         interface WithCertificate<ParentT> {
-            WithCertificate<ParentT> withPfxCertificateToUpload(File pfxFile, String password);
+            /**
+             * Upload a PFX certificate.
```
Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/HostNameSslBinding.java]: 3rd person declarative mood --> "Uploads ..."

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/HostNameSslBinding.java [hunk 6]
```diff
@@ -35,78 +36,230 @@
...
+             * @param pfxFile the PFX certificate file to upload
+             * @param password the password to the certificate
+             * @return the next stage of the hostname SSL binding definition
+             */
+            WithSslType<ParentT> withPfxCertificateToUpload(File pfxFile, String password);
+
+            /**
+             * Place a new App Service certificate order to use for the hostname
```
Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/HostNameSslBinding.java]: "Places ..."

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/HostNameSslBinding.java [hunk 7]
```diff
@@ -35,78 +36,230 @@
...
+            WithSslType<ParentT> withReadyToUseAppServiceCertificateOrder(AppServiceCertificateOrder certificateOrder);
         }
 
+        /**
+         * The stage of hostname SSL binding definition allowing key vault for certificate store to be specified.
+         * @param <ParentT> the return type of the final {@link WithAttach#attach()}
+         */
+        interface WithKeyVault<ParentT> {
+            /**
+             * Store the certificate in an existing vault.
```
Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/HostNameSslBinding.java]: "Stores..."

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/AppServiceCertificate.java (hunk 1 above)
Comment (jianghaolu) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/AppServiceCertificate.java]: Cer is not certificate here: https://en.wikipedia.org/wiki/Certified_Emission_Reduction

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/HostNameSslBinding.java [hunk 8]
```diff
@@ -35,78 +36,230 @@
...
+             * @param vault the existing vault to use
+             * @return the next stage of the hostname SSL binding definition
+             */
+            WithSslType<ParentT> withExistingKeyVault(Vault vault);
+
+            /**
+             * Create a new key vault and store the certificate in it.
```
Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/HostNameSslBinding.java]: "Creates... " and "stores..."

Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/HostNameSslBinding.java]: how about "Creates a new key vault to store the certificate in."

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/HostNameSslBinding.java [hunk 9]
```diff
@@ -35,78 +36,230 @@
...
+             * @param vaultName the name of the key vault to create
+             * @return the next stage of the hostname SSL binding definition
+             */
+            WithSslType<ParentT> withNewKeyVault(String vaultName);
+        }
+
+        /**
+         * The stage of hostname SSL binding definition allowing SSL type to be specified.
```
Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/HostNameSslBinding.java]: "The stage of a hostname SSL binding definition allowing the SSL type to be specified."

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/WebAppBase.java [hunk 10]
```diff
@@ -0,0 +1,792 @@
...
+     */
+    String autoSwapSlotName();
+
+    /**
+     * Start the web app or deployment slot.
+     */
+    void start();
+
+    /**
+     * Stop the web app or deployment slot.
```
Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/WebAppBase.java]: "Stops..."

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/WebAppBase.java [hunk 11]
```diff
@@ -0,0 +1,792 @@
...
+     */
+    void stop();
+
+    /**
+     * Restart the web app or deployment slot.
```
Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/WebAppBase.java]: "Restarts..."

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/WebAppBase.java [hunk 12]
```diff
@@ -0,0 +1,792 @@
...
+     * Grouping of all the site definition stages.
+     */
+    interface DefinitionStages {
+        /**
+         * A web app definition allowing app service plan to be set.
+         * @param <FluentT> the type of the resource, either a web app or a deployment slot
+         */
+        interface WithAppServicePlan<FluentT> {
+            /**
+             * Create a new free app service plan to use. No custom domains or SSL bindings are available in this plan.
```
Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/WebAppBase.java]: "Creates..."

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/WebAppBase.java [hunk 13]
```diff
@@ -0,0 +1,792 @@
...
+            /**
+             * Create a new app service plan to use.
+             * @param name the name of the app service plan
+             * @param pricingTier the pricing tier to use
+             * @return the next stage of web app definition
+             */
+            WithHostNameBinding<FluentT> withNewAppServicePlan(String name, AppServicePricingTier pricingTier);
+
+            /**
+             * Use an existing app service plan for the web app.
```
Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/WebAppBase.java]: "Uses.."
I won't comment on more, just recommend re-reading all javadocs and correcting to 3rd party declarative mood.

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/WebAppBase.java [hunk 14]
```diff
@@ -0,0 +1,792 @@
...
+
+            /**
+             * Specifies the managed pipeline mode.
+             * @param managedPipelineMode managed pipeline mode
+             * @return the next stage of web app definition
+             */
+            WithCreate<FluentT> withManagedPipelineMode(ManagedPipelineMode managedPipelineMode);
+
+            /**
+             * Specifies slot name to auto-swap when a deployment is completed in this web app / deployment slot.
```
Comment (martinsawicki) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/WebAppBase.java]: "Specifies the slot name..."

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/WebAppBase.java [hunk 15]
```diff
@@ -0,0 +1,792 @@
...
+
+    /**
+     * @return host names for the web app that are enabled
+     */
+    Set<String> enabledHostNames();
+
+    /**
+     * @return management information availability state for the web app.
+     *
+     * Normal means that the site is running correctly and that
```
Comment (jianghaolu) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/WebAppBase.java]: For auto-generated enums this is the only way right now - Swagger doesn't allow documentations for enums.

**Comment by martinsawicki** [comment]
Stil better to not list the constants on the method. They’re obvious when people see the enum. Someday these enums will have other constants. Not great for maintenance.

File: azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/AppServiceCertificate.java (hunk 1 above)
Comment (jianghaolu) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/AppServiceCertificate.java]: Oops what is that

Comment (jianghaolu) [inline azure-mgmt-website/src/main/java/com/microsoft/azure/management/website/AppServiceCertificate.java]: okay my point is cer file is like pfx file, it's a well-known file extension

**Pull request merged**